lib/browser/api/menu.js

```diff
@@ -272,71 +125,154 @@ Menu.prototype.insert = function (pos, item) {
   } else {
-    BrowserWindow.getAllWindows().forEach(function (window) {
-      window.setMenu(menu)
-    })
+    const windows = BrowserWindow.getAllWindows()
+    return windows.map(w => w.setMenu(menu))
   }
 }
 
-Menu.getApplicationMenu = function () {
-  return applicationMenu
+// build a menu by passing in a template
+Menu.buildFromTemplate = function (template) {
+  if (!(template instanceof Array)) {
+    throw new TypeError('Invalid template for Menu')
+  }
+
+  const menu = new Menu()
+  const positioned = []
+  let idx = 0
+
+  // sort template by position
+  template.forEach(item => {
+    idx = (item.position) ? indexToInsertByPosition(positioned, item.position) : idx += 1
+    positioned.splice(idx, 0, item)
+  })
+
+  // add each item from positioned menu to application menu
+  positioned.forEach((item) => {
+    if (typeof item !== 'object') {
+      throw new TypeError('Invalid template for MenuItem')
+    }
+    menu.append(new MenuItem(item))
+  })
+
+  return menu
 }
 
-Menu.sendActionToFirstResponder = bindings.sendActionToFirstResponder
+/* Helper Functions */
 
-Menu.buildFromTemplate = function (template) {
-  var insertIndex, item, j, k, len, len1, menu, menuItem, positionedTemplate
-  if (!Array.isArray(template)) {
-    throw new TypeError('Invalid template for Menu')
+// Search between separators to find a radio menu item and return its group id
+function generateGroupId (items, pos) {
+  if (pos > 0) {
+    for (let idx = pos - 1; idx >= 0; idx--) {
+      if (items[idx].type === 'radio') return items[idx].groupId
+      if (items[idx].type === 'separator') break
+    }
+  } else if (pos < items.length) {
+    for (let idx = pos; idx <= items.length - 1; idx++) {
+      if (items[idx].type === 'radio') return items[idx].groupId
+      if (items[idx].type === 'separator') break
+    }
+  }
+  groupIdIndex += 1
+  return groupIdIndex
+}
+
+// Returns the index of item according to |id|.
+function indexOfItemById (items, id) {
+  const foundItem = items.find(item => item.id === id) || -1
+  return items.indexOf(foundItem)
+}
+
+// Returns the index of where to insert the item according to |position|
+function indexToInsertByPosition (items, position) {
+  if (!position) return items.length
+
+  const [query, id] = position.split('=') // parse query and id from position
+  const idx = indexOfItemById(items, id) // calculate initial index of item
+
+  // warn if query doesn't exist
+  if (idx === -1 && query !== 'endof') {
+    console.warn(`Item with id ${id} is not found`)
+    return items.length
   }
-  positionedTemplate = []
-  insertIndex = 0
-  for (j = 0, len = template.length; j < len; j++) {
-    item = template[j]
-    if (item.position) {
-      insertIndex = indexToInsertByPosition(positionedTemplate, item.position)
-    } else {
-      // If no |position| is specified, insert after last item.
-      insertIndex++
+
+  // compute new index based on query
+  const queries = {
+    'after': (index) => {
```

Comment:
why quotes? also, could use short es6 method notaion?

Reply:
quotes will be removed, whoops. I didn't use `(index) => { index += 1 }`, because our linter disallows returning assignment

Reply:
Just `Index + 1` ?